File: apps/web/src/services/aiRagCopilot.ts
import { apiClient } from './apiClient';

export interface KnowledgeCollection {
  id: string;
  organization_id: string;
  name: string;
  description?: string;
  visibility: string;
  status: string;
  created_at: string;
}

export interface RAGDocument {
  id: string;
  collection_id: string;
  document_name: string;
  file_type: string;
  file_size: number;
  status: string;
  created_at: string;
}

export interface RAGChatMessage {
  id: string;
  session_id: string;
  role: string;
  message: string;
  tokens: number;
  latency: number;
  created_at: string;
}

export interface RAGChatSession {
  id: string;
  organization_id: string;
  session_name: string;
  model_name: string;
  temperature: number;
  status: string;
  created_at: string;
  messages: RAGChatMessage[];
}

export interface PromptTemplate {
  id: string;
  name: string;
  category: string;
  description?: string;
  system_prompt: string;
  status: string;
  created_at: string;
}

export interface AIAgent {
  id: string;
  organization_id: string;
  agent_name: string;
  agent_type: string;
  system_prompt: string;
  model: string;
  temperature: number;
  status: string;
  created_at: string;
}

export interface ToolRegistry {
  id: string;
  tool_name: string;
  tool_description: string;
  tool_type: string;
  endpoint?: string;
  enabled: boolean;
  created_at: string;
}

export interface AIDashboardSummary {
  total_documents: number;
  total_embeddings: number;
  total_collections: number;
  active_chat_sessions: number;
  total_agent_runs: number;
  average_response_time_sec: number;
  total_prompt_templates: number;
  total_token_usage: number;
}

export const aiRagCopilotService = {
  // Dashboard
  getDashboardSummary: async (orgId: string): Promise<AIDashboardSummary> => {
    const res = await apiClient.get('/api/v1/ai/dashboard', { params: { org_id: orgId } });
    return res.data;
  },

  // Collections
  getCollections: async (orgId: string): Promise<KnowledgeCollection[]> => {
    const res = await apiClient.get('/api/v1/ai/collections', { params: { org_id: orgId } });
    return res.data;
  },

  createCollection: async (data: Partial<KnowledgeCollection>): Promise<KnowledgeCollection> => {
    const res = await apiClient.post('/api/v1/ai/collections', data);
    return res.data;
  },

  // Documents
  getDocuments: async (collectionId?: string): Promise<RAGDocument[]> => {
    const res = await apiClient.get('/api/v1/ai/documents', { params: { collection_id: collectionId } });
    return res.data;
  },

  ingestDocument: async (data: any): Promise<RAGDocument> => {
    const res = await apiClient.post('/api/v1/ai/documents', data);
    return res.data;
  },

  // Copilot Chat
  getChatSessions: async (orgId: string): Promise<RAGChatSession[]> => {
    const res = await apiClient.get('/api/v1/ai/copilot/sessions', { params: { org_id: orgId } });
    return res.data;
  },

  createChatSession: async (data: any): Promise<RAGChatSession> => {
    const res = await apiClient.post('/api/v1/ai/copilot/sessions', data);
    return res.data;
  },

  sendMessage: async (sessionId: string, message: string): Promise<RAGChatMessage> => {
    const res = await apiClient.post('/api/v1/ai/copilot/chat', {
      session_id: sessionId,
      role: 'user',
      message: message,
    });
    return res.data;
  },

  // Prompts
  getPromptTemplates: async (): Promise<PromptTemplate[]> => {
    const res = await apiClient.get('/api/v1/ai/prompts');
    return res.data;
  },

  createPromptTemplate: async (data: any): Promise<PromptTemplate> => {
    const res = await apiClient.post('/api/v1/ai/prompts', data);
    return res.data;
  },

  // Agents
  getAgents: async (orgId: string): Promise<AIAgent[]> => {
    const res = await apiClient.get('/api/v1/ai/agents', { params: { org_id: orgId } });
    return res.data;
  },

  createAgent: async (data: any): Promise<AIAgent> => {
    const res = await apiClient.post('/api/v1/ai/agents', data);
    return res.data;
  },

  runAgent: async (agentId: string, inputText: string): Promise<any> => {
    const res = await apiClient.post('/api/v1/ai/agents/run', {
      agent_id: agentId,
      input_text: inputText,
    });
    return res.data;
  },

  // Tools
  getTools: async (): Promise<ToolRegistry[]> => {
    const res = await apiClient.get('/api/v1/ai/tools');
    return res.data;
  },
};

File: apps/api/app/schemas/ai_rag_copilot.py
import uuid
from datetime import datetime
from pydantic import BaseModel, ConfigDict, Field


# --- Knowledge Base Schemas ---
class KnowledgeCollectionBase(BaseModel):
    name: str = Field(..., min_length=1, max_length=255)
    description: str | None = Field(None, max_length=1000)
    visibility: str = Field(default="Internal", max_length=50)
    status: str = Field(default="Active", max_length=50)


class KnowledgeCollectionCreate(KnowledgeCollectionBase):
    organization_id: uuid.UUID


class KnowledgeCollectionResponse(KnowledgeCollectionBase):
    id: uuid.UUID
    organization_id: uuid.UUID
    created_at: datetime
    updated_at: datetime

    model_config = ConfigDict(from_attributes=True)


class RAGDocumentBase(BaseModel):
    document_name: str = Field(..., min_length=1, max_length=255)
    file_type: str = Field(..., min_length=1, max_length=50)  # pdf, docx, txt, xlsx
    file_size: int = Field(default=0, ge=0)
    status: str = Field(default="Processed", max_length=50)


class RAGDocumentCreate(RAGDocumentBase):
    collection_id: uuid.UUID
    document_content: str = Field(..., min_length=1)  # document text to ingest


class RAGDocumentResponse(RAGDocumentBase):
    id: uuid.UUID
    collection_id: uuid.UUID
    uploaded_by: uuid.UUID | None = None
    created_at: datetime
    updated_at: datetime

    model_config = ConfigDict(from_attributes=True)


# --- Copilot Chat Schemas ---
class RAGChatMessageBase(BaseModel):
    role: str = Field(..., max_length=50)  # user, assistant, system
    message: str = Field(..., min_length=1)


class RAGChatMessageCreate(RAGChatMessageBase):
    session_id: uuid.UUID


class RAGChatMessageResponse(RAGChatMessageBase):
    id: uuid.UUID
    session_id: uuid.UUID
    tokens: int
    latency: float
    created_at: datetime
    updated_at: datetime

    model_config = ConfigDict(from_attributes=True)


class RAGChatSessionBase(BaseModel):
    session_name: str = Field(..., min_length=1, max_length=255)
    model_name: str = Field(default="gpt-4o", max_length=100)
    temperature: float = Field(default=0.7, ge=0.0, le=2.0)
    status: str = Field(default="Active", max_length=50)


class RAGChatSessionCreate(RAGChatSessionBase):
    organization_id: uuid.UUID
    user_id: uuid.UUID | None = None


class RAGChatSessionResponse(RAGChatSessionBase):
    id: uuid.UUID
    organization_id: uuid.UUID
    user_id: uuid.UUID | None = None
    created_at: datetime
    updated_at: datetime
    messages: list[RAGChatMessageResponse] = Field(default_factory=list)

    model_config = ConfigDict(from_attributes=True)


# --- Prompt Engineering Schemas ---
class PromptTemplateBase(BaseModel):
    name: str = Field(..., min_length=1, max_length=255)
    category: str = Field(default="General", max_length=100)
    description: str | None = Field(None, max_length=1000)
    system_prompt: str = Field(..., min_length=1)
    status: str = Field(default="Active", max_length=50)


class PromptTemplateCreate(PromptTemplateBase):
    pass


class PromptTemplateResponse(PromptTemplateBase):
    id: uuid.UUID
    created_at: datetime
    updated_at: datetime

    model_config = ConfigDict(from_attributes=True)


# --- AI Agents & Tool Calling Schemas ---
class ToolRegistryBase(BaseModel):
    tool_name: str = Field(..., min_length=1, max_length=100)
    tool_description: str = Field(..., min_length=1, max_length=1000)
    tool_type: str = Field(default="API", max_length=50)
    endpoint: str | None = Field(None, max_length=500)
    enabled: bool = Field(default=True)


class ToolRegistryCreate(ToolRegistryBase):
    pass


class ToolRegistryResponse(ToolRegistryBase):
    id: uuid.UUID
    created_at: datetime
    updated_at: datetime

    model_config = ConfigDict(from_attributes=True)


class AIAgentBase(BaseModel):
    agent_name: str = Field(..., min_length=1, max_length=255)
    agent_type: str = Field(default="Conversational", max_length=100)
    system_prompt: str = Field(..., min_length=1)
    model: str = Field(default="gpt-4o", max_length=100)
    temperature: float = Field(default=0.7, ge=0.0, le=2.0)
    status: str = Field(default="Active", max_length=50)


class AIAgentCreate(AIAgentBase):
    organization_id: uuid.UUID


class AIAgentResponse(AIAgentBase):
    id: uuid.UUID
    organization_id: uuid.UUID
    created_at: datetime
    updated_at: datetime

    model_config = ConfigDict(from_attributes=True)


class AgentRunCreate(BaseModel):
    agent_id: uuid.UUID
    input_text: str = Field(..., min_length=1)


class AgentRunResponse(BaseModel):
    id: uuid.UUID
    agent_id: uuid.UUID
    input_text: str
    output_text: str
    execution_time: float
    status: str
    created_at: datetime
    updated_at: datetime

    model_config = ConfigDict(from_attributes=True)


# --- AI Dashboard Summary Schema ---
class AIDashboardSummary(BaseModel):
    total_documents: int
    total_embeddings: int
    total_collections: int
    active_chat_sessions: int
    total_agent_runs: int
    average_response_time_sec: float
    total_prompt_templates: int
    total_token_usage: int
